src/main/java/seedu/duke/Duke.java

```diff
@@ -0,0 +1,784 @@
+            }
+        } catch (IOException e) {
+            System.out.println("An error occurred while saving tasks to file: " + e.getMessage());
+        }
+    }
+    /**
+     * Converts a task to a formatted string for saving to the file.
+     * The format includes the task type, status, description, and any additional information.
+     *
+     * @param task The task to be converted to a string.
+     * @return A formatted string representing the task.
+     */
+    private String taskToFileString(Task task) {
+        String type = task instanceof Todo ? "T" :
+                task instanceof Deadline ? "D" :
+                        task instanceof Event ? "E" : "";
+        String status = task.isDone ? "1" : "0";
+        String details = task.getDescription();
+        String additionalInfo = "";
+
+        if (task instanceof Deadline) {
+            Deadline deadline = (Deadline) task;
+            additionalInfo = " | " + deadline.getBy().format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm"));
+        } else if (task instanceof Event) {
+            Event event = (Event) task;
+            additionalInfo = " | " + event.getFrom().format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm")) +
+                    " | " + event.getTo().format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm"));
+        }
+
+        return type + " | " + status + " | " + details + additionalInfo;
+    }
+
+}
+
+/**
+ * Handles parsing of user input commands and converting them into Command objects.
+ */
+class Parser {
+    private static final DateTimeFormatter dateTimeFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HHmm");
+    /**
+     * Parses user input and returns the appropriate Command object.
+     *
+     * @param fullCommand The full user input command.
+     * @return The command object corresponding to the user input.
+     * @throws DukeException If the command is invalid or if the input format is incorrect.
+     */
+    public static Command parse(String fullCommand) throws DukeException {
+        String[] commandParts = fullCommand.split(" ", 2);
+        String commandType = commandParts[0];
+        String commandArgs = commandParts.length > 1 ? commandParts[1] : "";
+
+        switch (commandType) {
+        case "todo":
+            if (commandArgs.isEmpty()) {
+                throw new DukeException("The description of a todo cannot be empty.");
+            }
+            return new AddTodoCommand(commandArgs);
+        case "deadline":
+            return parseAddDeadlineCommand(commandArgs);
+        case "event":
+            return parseAddEventCommand(commandArgs);
+        case "list":
+            return new ListCommand();
+        case "mark":
+            try {
+                int index = Integer.parseInt(commandArgs) - 1;
+                return new MarkCommand(index);
+            } catch (NumberFormatException e) {
+                throw new DukeException("Invalid task number format.");
```

Comment:
Would there be less duplication if a static final variable was used to store the string here? Consider the case that the developer decides to make minor changes to this string, if the string was stored "INVALID_TASK_NUMBER = ...", and all other codes reference that, at least only that constant needs to be edited rather than all other instances where it was used 😃 

I noticed the same for other strings, where there were also duplicates